File: Lab6/Prog2_Lab6/Prog2_Lab6_C++_Lib/Expression.cpp
#include "pch.h"
#include "Expression.h"

// Метод для перевірки правильності введених операнд, виконується при введені даних та розрахунку значення виразу.
void Expression::Check()
{
    try {
        // Перевірка кореня
        if (isnan(sqrt(41 - d))) { throw "Квадратний корiнь вiд'ємного числа"; }
        // Перевірка дільника
        else if ((sqrt(41 - d) - (b * a) + c) == 0) { throw "Дiльник дорiвнює нулю"; }
        return;
    }
    catch (char* errMessage) {
        throw std::runtime_error(errMessage);
    }
}

// Метод для розрахунку значення виразу
double Expression::Calculate()
{
    // Викликається перевірка правильності операнд
    Check();
    // Розраховується значення виразу
    return ((a * b / 4) - 1) / (sqrt(41 - d) - (b * a) + c);
}

// Конструктор
Expression::Expression(double aInput, double bInput, double cInput, double dInput)
{
    a = aInput;
    b = bInput;
    c = cInput;
    d = dInput;

    // Викликається перевірка правильності операнд
    Check();
}

// Індексатор для введення-виведення операнд виразу
// Перевіряє правильність операнд при введенні
void Expression::Change(char index, double value)
{
    try {
        switch (index) {
        case 'a': a = value; break;
        case 'b': b = value; break;
        case 'c': c = value; break;
        case 'd': d = value; break;
        default: throw "Символ не відповідає жодному з операнд виразу."; break;
        }
    }
    catch (char* errMessage) {
        throw std::invalid_argument(errMessage);
    }

    // Викликається перевірка правильності операнд
    Check();
}

File: Lab6/Prog2_Lab6/Prog2_Lab6_Cs_Lib/Expression.cs
using System;

namespace Prog2_Lab6_Cs_Lib
{
    public class Expression
    {
        private double a;
        private double b;
        private double c;
        private double d;

        // Метод для перевірки правильності введених операнд, виконується при введені даних та розрахунку значення виразу.
        private void Check()
        {
            // Перевірка кореня
            if (Double.IsNaN(Math.Sqrt(41 - d))) { throw new ArithmeticException("Квадратний корінь від'ємного числа"); }
            // Перевірка дільника
            else if ((Math.Sqrt(41 - d) - (b * a) + c) == 0) { throw new ArithmeticException("Дільник дорівнює нулю"); }
            return;
        }

        // Метод для розрахунку значення виразу
        public double Calculate()
        { 
            // Викликається перевірка правильності операнд
            Check();
            // Розраховується значення виразу
            return((a * b / 4) - 1) / (Math.Sqrt(41 - d) - (b * a) + c);
        }

        // Конструктор
        public Expression(double aInput, double bInput, double cInput, double dInput)
        {
            a = aInput;
            b = bInput;
            c = cInput;
            d = dInput;

            // Викликається перевірка правильності операнд
            Check();
        }

        // Індексатор для введення-виведення операнд виразу
        // Перевіряє правильність операнд при введенні
        public double this[char index]
        {
            get
            {
                if (index == 'a') { return a; }
                else if (index == 'b') { return b; }
                else if (index == 'c') { return c; }
                else if (index == 'd') { return d; }
                else { throw new ArgumentException("Символ не відповідає жодному з операнд виразу."); }
            }
            set
            {
                if (index == 'a') { a = value; }
                else if (index == 'b') { b = value; }
                else if (index == 'c') { c = value; }
                else if (index == 'd') { d = value; }
                else { throw new ArgumentException("Символ не відповідає жодному з операнд виразу."); }

                // Викликається перевірка правильності операнд
                Check();
            }
        }
    }
}
